source/_posts/2017-08-16-Getting-Stack-Traces-in-Swift-on-Linux.md

```diff
@@ -0,0 +1,125 @@
+...
+0000000100001160 t __T07example4UserVwxx
+```
+
+> If you wish to learn more about mangling in programming languages, check out [this](https://en.wikipedia.org/wiki/Name_mangling) wikipedia article.
+
+The process of extracting metadata from a mangled symbol is known as demangling ([Swift Name Demangling](https://mikeash.com/pyblog/friday-qa-2014-08-15-swift-name-mangling.html)). The way Swift mangles symbols is not well documented and is subject to change across versions of Swift. Matt Gallagher [translated](https://github.com/mattgallagher/CwlDemangle) Swift's internal version of the mangler to Swift and open-sourced it. This tool works quite well for Swift 3, but Swift 4 changed the way most of the mangling works. For Swift 4 support we moved to the internal implementation using a [little linker hack](https://github.com/nodes-vapor/stacked/pull/12) (thanks [@harlanhaskins](https://twitter.com/harlanhaskins)). This implementation is quite brittle, but, until we get a version that is exposed to Swift, it is our only option.
+
+Now, when we try to run again, we are faced with another issue.
+
+## Swift symbols not being exported on Linux
+
+By default, dynamic symbols are not exported on Linux. That means that all Swift functions in the stack trace will not have their name displayed, just like in the table below.
+
+|     | Symbol                                                 | Address        |
+| --- | -------------------------------------------------------| -------------- |
+| 0   | .build/release/libFrameAddress.so(get_stack_trace+0x2d)| 0x7f03ad5ca6ad |
+| 1   | .build/release/App()                                   | 0x7f17b0       |
+| 2   | .build/release/App()                                   | 0x7fb3fa       |
+| ... |              
+
+```c
+libFrameAddress.so(get_stack_trace+0x2d)
+```
+
+`libFrameAddress.so` is a C binary and `get_stack_trace` is our stack trace implementation.
+
+```
+App()
+```
+
+`App` is the Swift binary and the name and address of the function are missing. In order to export this information, we need to pass the following flag to the linker:
+
+`-Xlinker --export-dynamic`
+
+We now have our beautiful stack traces!
+
+| | Symbol |
+| --- | --- |
+| 0 | get_stack_trace |
+| 1 | static Stacked.FrameAddress.getStackTrace(maxStackSize: Swift.Int) -> Swift.Array<Swift.String> |
+| 2 | StackedTests.StackedTests.anotherExample() -> Swift.Array<Swift.String> |
+| 3 | StackedTests.StackedTests.testExample() -> () |
+| 4 | @objc StackedTests.StackedTests.testExample() -> () |
+
+## Consclusion
+
+Getting stack traces in Linux was much more involved than we had expected. We ended up touching nearly every part of the system and were plagued with issues. That being said, it was worth the effort and will greatly ease debugging. Hopefully Foundation and Vapor will add native support for stack traces and remove the burden of having to install a 3rd party repository and passing in build flags on Linux. 
```

Comment:
Maybe a quick sentence or two about how one would use this? And perhaps end the post with links to all the relevant repos and what they do (if people are interested and might want to contribute).